python/cudf/cudf/tests/utils.py

```diff
@@ -124,6 +124,8 @@ def assert_exceptions_equal(
     lfunc_args_and_kwargs=None,
     rfunc_args_and_kwargs=None,
     compare_error_message=True,
+    expected_exception=None,
+    expected_error_message=None,
```

Comment:
Are these options provided to account for when the `rfunc` does not raise the same type of exception as `lfunc`? If so, I'd say the intent is better clarified with something like this:

```suggestion
    check_exception_type=True,
    check_exception_message=True,
```

My opinion is that if the exception type of `lfunc` and `rfunc` don't match, that's not a very useful comparison to make. So we should probably not even have a `check_exception_type` message. Are there any situations in which we raise a different class of exception from Pandas and test for that?

Reply:
>  Are there any situations in which we raise a different class of exception from Pandas and test for that?

Yes, there are multiple places this happens where pandas has their custom Error class defined or either call a numpy api where numpy usually has their own exception types:

https://github.com/rapidsai/cudf/blob/bff8e854823adea5d6a0e86f1d38cb2c1a507818/python/cudf/cudf/tests/test_categorical.py#L159-L172


https://github.com/rapidsai/cudf/blob/bff8e854823adea5d6a0e86f1d38cb2c1a507818/python/cudf/cudf/tests/test_categorical.py#L236-L253

https://github.com/rapidsai/cudf/blob/bff8e854823adea5d6a0e86f1d38cb2c1a507818/python/cudf/cudf/tests/test_joining.py#L456-L468

https://github.com/rapidsai/cudf/blob/bff8e854823adea5d6a0e86f1d38cb2c1a507818/python/cudf/cudf/tests/test_timedelta.py#L338-L346

another example where we raise `TypeError` and pandas calls numpy which has their own custom error:
```
>>> td = pd.Series([1, 2, 3], dtype='timedelta64[ns]')
>>> td * td
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/Users/pgali/PycharmProjects/del/venv1/lib/python3.7/site-packages/pandas/core/ops/common.py", line 65, in new_method
    return method(self, other)
  File "/Users/pgali/PycharmProjects/del/venv1/lib/python3.7/site-packages/pandas/core/ops/__init__.py", line 343, in wrapper
    result = arithmetic_op(lvalues, rvalues, op)
  File "/Users/pgali/PycharmProjects/del/venv1/lib/python3.7/site-packages/pandas/core/ops/array_ops.py", line 186, in arithmetic_op
    res_values = op(lvalues, rvalues)
  File "/Users/pgali/PycharmProjects/del/venv1/lib/python3.7/site-packages/pandas/core/ops/common.py", line 65, in new_method
    return method(self, other)
  File "/Users/pgali/PycharmProjects/del/venv1/lib/python3.7/site-packages/pandas/core/arrays/timedeltas.py", line 478, in __mul__
    result = self._data * other
  File "/Users/pgali/PycharmProjects/del/venv1/lib/python3.7/site-packages/pandas/core/ops/common.py", line 65, in new_method
    return method(self, other)
  File "/Users/pgali/PycharmProjects/del/venv1/lib/python3.7/site-packages/pandas/core/arrays/timedeltas.py", line 478, in __mul__
    result = self._data * other
numpy.core._exceptions.UFuncTypeError: ufunc 'multiply' cannot use operands with types dtype('<m8[ns]') and dtype('<m8[ns]')
```

Reply:
Thanks -- that makes sense. I'm in favor of `check_exception_type=True` rather than `expected_exception_type=None`.

Firstly, it is consistent with the arguments to the other `assert_*` functions we have, e.g., `check_dtype`, `check_index`, etc.,

Secondly, it preserves symmetry, so it doesn't matter if the `lfunc` or `rfunc` is the test target:

```
assert_exceptions_equal(cudf_func, pandas_func, ..., check_exception_type=False)
```

In terms of implementation, if `check_exception_type=False`, we can simply verify that both the `lfunc` and `rfunc` raise an exception of the base type `Exception`.

Reply:
Okay, in addition can we also have `expected_exception` & `expected_error_message` which are the fallback assertion values? - Asking this because if we remove these above two parameters, we might have to re-introduce the old assertion for the `pytests` where we want to assert for the exact exception type.

Signature would look something like this..
```python
def assert_exceptions_equal(
    lfunc,
    rfunc,
    lfunc_args_and_kwargs=None,
    rfunc_args_and_kwargs=None,
    check_exception_type=True,
    check_error_message=True,
    expected_exception_type=None,
    expected_error_message=None,
)
```

Reply:
Hmm, I'd definitely like to avoid complicating the API if possible.

If we want to assert that a piece of code raises a specific type of exception, we should use `pytest.raises`:

```python
with pytest.raises(ExceptionType):
     ...
```

If we want to assert that two function calls raise the **same** exception, then we should use `assert_exceptions_equal`:

```python
assert_exceptions_equal(func_1, func_2, func_1_args, func_2_args, ...)
```

I don't think  we should try to do both with `assert_exceptions_equal`.

For the cases you mentioned above, I think it's fair to say that Pandas/NumPy have fundamentally different behaviour from cuDF. Thus, we shouldn't be comparing their behaviour with ours, but rather just verifying our own behaviour.

Reply:
Discussed offline and we decided on:

```
def assert_exceptions_equal(
    lfunc,
    rfunc,
    lfunc_args_and_kwargs=None,
    rfunc_args_and_kwargs=None,
    check_exception_type=True,
    check_error_message=True,
    expected_error_message=None,
)
```

Reply:
Changed to the above function signature as discussed offline 👍